src/EditorFeatures/Core/ExternalAccess/IntelliCode/IntentProcessor.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ExternalAccess.IntelliCode.Api;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.IntelliCode
+{
+    [Export(typeof(IIntentProcessor)), Shared]
+    internal class IntentProcessor : IIntentProcessor
+    {
+        private readonly ImmutableDictionary<(string LanguageName, string IntentName), Lazy<IIntentProvider, IIntentProviderMetadata>> _lazyIntentProviders;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public IntentProcessor([ImportMany] IEnumerable<Lazy<IIntentProvider, IIntentProviderMetadata>> lazyIntentProviders)
+        {
+            _lazyIntentProviders = CreateProviderMap(lazyIntentProviders);
+        }
+
+        private static ImmutableDictionary<(string LanguageName, string IntentName), Lazy<IIntentProvider, IIntentProviderMetadata>> CreateProviderMap(
+            IEnumerable<Lazy<IIntentProvider, IIntentProviderMetadata>> lazyIntentProviders)
+        {
+            return lazyIntentProviders.ToImmutableDictionary(
+                provider => (provider.Metadata.LanguageName, provider.Metadata.IntentName),
+                provider => provider);
+        }
+
+        public async Task<ImmutableArray<IntentResult>> ComputeEditsAsync(IntentRequestContext intentRequestContext, CancellationToken cancellationToken)
+        {
+            var currentDocument = intentRequestContext.CurrentSnapshotSpan.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            Contract.ThrowIfNull(currentDocument);
```

Comment:
i woudl say this shoudl either be a real throw, or we bail out gracefully.  across module boundaries it's less clear what the contract should be.  for now, a contract call is ok as we test out with them.

Reply:
the benefit of being an IIntentSource and IIntentSourceProvider is that we will export ourselves appropraitely with mef content tyeps and whatnot.